### PR TITLE
[FrameworkBundle] Deprecate making `cache.app` adapter taggable

### LINKED ISSUE #54339: Forbid making cache.app tag-aware
### Description

Follow-up from https://github.com/twigphp/Twig/issues/3869 

Since `app.cache.tagware` exists by default, we should not allow `app.cache` to be taggable. Whenever people try to configure `app.cache` as taggable we should inform them about `app.cache.taggable`.

Ideally, in the current version this would raise a deprecation notice, and it should be completely forbidden in the next Symfony version.

### Example

_No response_

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Issues        | Fix #54339
| License       | MIT

Feedback appreciated on the deprecation message

### CODE REVIEW

File: UPGRADE-7.2.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Cache
 -----
 
  * `igbinary_serialize()` is not used by default when the igbinary extension is installed
+ * Deprecate making `cache.app` adapter taggable, use `cache.app.taggable` adapter instead
```
Comment (xabbuh) [inline UPGRADE-7.2.md]:
```suggestion
 * Deprecate making `cache.app` adapter taggable, use the `cache.app.taggable` adapter instead
```

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * Make the `config/` directory optional in `MicroKernelTrait`, add support for service arguments in the
    invokable Kernel class, and register `FrameworkBundle` by default when the `bundles.php` file is missing
  * [BC BREAK] The `secrets:decrypt-to-local` command terminates with a non-zero exit code when a secret could not be read
+ * Deprecate making `cache.app` adapter taggable, use `cache.app.taggable` adapter instead
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]:
```suggestion
 * Deprecate making `cache.app` adapter taggable, use the `cache.app.taggable` adapter instead
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -2392,6 +2392,10 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
                 $container->setAlias('.'.$name.'.inner', $name);
                 $definition->addTag('cache.taggable', ['pool' => $name]);
             } elseif ($pool['tags']) {
+                if (\in_array('cache.app', $pool['adapters'], true)) {
+                    trigger_deprecation('symfony/framework-bundle', '7.2', 'Using the "tags" option with the the default "cache.app" adapter is deprecated. Either use the "cache.app.taggable" adapter or give another tag aware adapter explicitly. The "cache.app.taggable" adapter will be use by default when the "tags" option is used in Symfony 8.0.');
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
                    trigger_deprecation('symfony/framework-bundle', '7.2', 'Using the "tags" option with the the default "cache.app" adapter is deprecated. Either use the "cache.app.taggable" adapter or give another tag aware adapter explicitly. The "cache.app.taggable" adapter will be used by default when the "tags" option is used in Symfony 8.0.');
```

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/cache_cacheapp_tagaware.xml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:framework="http://symfony.com/schema/dic/symfony"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd
+        http://symfony.com/schema/dic/symfony https://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+    <framework:config http-method-override="false" handle-all-throwables="true">
+        <framework:annotations enabled="false" />
+        <framework:php-errors log="true" />
+        <framework:cache>
+            <framework:pool name="app.cache" adapter="cache.app" tags="true" />
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/cache_cacheapp_tagaware.xml]:
```suggestion
            <framework:pool name="app.tagaware" adapter="cache.app" tags="true" />
```

**Comment by alexandre-daubois** [comment]
Thanks, comments addressed :+1: I'm having a look at related failures

**Comment by alexandre-daubois** [comment]
CI is green now

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -2363,6 +2363,10 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
             ];
         }
         foreach ($config['pools'] as $name => $pool) {
+            if (\in_array('cache.app', $pool['adapters'] ?? [], true) && $pool['tags']) {
+                trigger_deprecation('symfony/framework-bundle', '7.2', 'Using the "tags" option with the "cache.app" adapter is deprecated. You should either use the "cache.app.taggable" adapter or give another tag aware adapter explicitly to the "tags" option. Using the "cache.app" adapter with the "tags" option will throw an exception in Symfony 8.0.');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: let's add the code throwing the exception in a comment

Comment (alexandre-daubois) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Here you go

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -2363,6 +2363,11 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
...
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]:
```suggestion
                trigger_deprecation('symfony/framework-bundle', '7.2', 'Using the "tags" option with the "cache.app" adapter is deprecated. You can use the "cache.app.taggable" adapter instead (aliased to the TagAwareCacheInterface for autowiring).');
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 7]
```diff
@@ -2363,6 +2363,11 @@ private function registerCacheConfiguration(array $config, ContainerBuilder $con
...
+                // throw new InvalidArgumentException('The "tags" option cannot be used with the "cache.app" adapter. You should use the "cache.app.taggable" adapter or give another tag aware adapter explicitly to the "tags" option.');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: we use LogicException

TagAwareCacheInterface
```suggestion
                // throw new LogicException('The "tags" option cannot be used with the "cache.app" adapter. You can use the "cache.app.taggable" adapter instead (aliased to the TagAwareCacheInterface for autowiring).');
```

Comment (alexandre-daubois) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Wording and exception class updated

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @alexandre-daubois.

**Pull request merged**